File: java/Easy/_680_ValidPalindromeII.java
package Easy;

import org.testng.Assert;
import org.testng.annotations.Test;

public class _680_ValidPalindromeII{
    public boolean validPalindrome(String s) {
        for(int i = 0; i<s.length()/2; i++){
            if(s.charAt(i) == s.charAt(s.length()-1-i)) continue;
            else{
                boolean left = isValid(s.substring(0,i) + s.substring(i+1));
                boolean right =isValid(s.substring(0,s.length()-1-i) + s.substring(s.length()-1-i+1));
                return left || right;
            }
        }
        return true;
    }

    private boolean isValid(String s){
        for(int i = 0; i<s.length()/2; i++)
            if(s.charAt(i) != s.charAt(s.length()-1-i)) return false;
        return true;
    }

    @Test
    public void test(){
        Assert.assertEquals( false, validPalindrome("abc"));
        Assert.assertEquals( true, validPalindrome("deeee"));
        Assert.assertEquals( true, validPalindrome("aba"));
        Assert.assertEquals( true, validPalindrome("abca"));
    }
}

File: Python3/_680_ValidPalindromeII.py
class Solution:
    def validPalindrome(self, s: str) -> bool:
        isPal = self.isPalindrome(s)
        if isPal:
            return True

        for i in range((int)(len(s)/2)):
            if s[i] != s[-i-1]:
                left = s[0:i]+s[i+1:]
                right = s[0:-i-1]+s[-i:]
                if i == 0:
                    right = s[0:-i-1]
                return self.isPalindrome(left) or self.isPalindrome(right)
        
    
    def isPalindrome(self, s: str) -> bool:
        for i in range((int)(len(s)/2)):
            if s[i] != s[-i-1]:
                return False
        return True

test = Solution()

print(test.validPalindrome("eccer"))
print(test.validPalindrome("abca"))
print(test.validPalindrome("abc"))
print(test.validPalindrome("abcddcba"))
